source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+// matrices that actually represent the transpose of what the Slang matrix
+// represented.
+// Both API's adopt the standard matrix multiplication convention whereby the
+// column count of the matrix on the left hand side needs to match row count of
+// the matrix on the right hand side.
+// For these reasons, and due to the fact that (M_1 ... M_n)^T = M_n^T ... M_1^T,
+// the order of matrix (and vector-matrix products) products must also reversed
+// in the WGSL code.
+//
+// This may lead to confusion (which is why this note is referenced in several
+// places), but the benefit of doing this is that the generated WGSL code is
+// simpler to generate and should be faster to compile.
+// A "terminology preserving" implementation would have to generate lots of
+// 'transpose' calls, or else perform more complicated transformations that
+// end up duplicating expressions many times.
+
+namespace Slang {
+
+void WGSLSourceEmitter::emitSwitchCaseSelectorsImpl(
+    IRBasicType *const switchConditionType,
+    SwitchRegion::Case const*const currentCase, bool const isDefault
+    )
+{
+    m_writer->emit("case ");
+    for (auto caseVal : currentCase->values)
+    {
+        if (caseVal->getOp() == kIROp_IntLit)
+        {
+            auto caseLitInst = static_cast<IRConstant*>(caseVal);
+            IRBasicType *const caseInstType {as<IRBasicType>(caseLitInst->getDataType())};
+            // WGSL doesn't allow switch condition and case type mismatches, see [1].
+            // Thus we need to insert explicit conversions.
+            // Doing a wrapping cast will match Slang's de facto semantics, according to
+            // [2].
+            // (This is just a bitcast, assuming a two's complement representation.)
+            // [1] https://www.w3.org/TR/WGSL/#switch-statement
+            // [2] https://github.com/shader-slang/slang/issues/4921
+            bool const needBitcast {
+                caseInstType->getBaseType() != switchConditionType->getBaseType()
+            };
+            if (needBitcast)
+            {
+                m_writer->emit("bitcast<");
+                emitType(switchConditionType);
+                m_writer->emit(">(");
+            }
+            emitOperand(caseVal, getInfo(EmitOp::General));
+            if (needBitcast)
+            {
+                m_writer->emit(")");
+            }
+        }
+        else
+        {
+            emitOperand(caseVal, getInfo(EmitOp::General));
+        }
+        m_writer->emit(", ");
+    }
+    if (isDefault)
+    {
+        m_writer->emit("default, ");
+    }
+    m_writer->emit(":\n");
+}
+
+void WGSLSourceEmitter::emitParameterGroupImpl(
+    IRGlobalParam* varDecl, IRUniformParameterGroupType* type
+)
+{
+    auto varLayout = getVarLayout(varDecl);
+    SLANG_RELEASE_ASSERT(varLayout);
+
+    for (auto attr : varLayout->getOffsetAttrs())
+    {
+
+        LayoutResourceKind const kind {attr->getResourceKind()};
```

Comment:
style:
LayoutResourceKind kind = attr->getResourceKind();

We don't use `const` if not necessary.